src/eins-network-id.c

```diff
@@ -0,0 +1,238 @@
+/* Copyright 2017 Endless Mobile, Inc.
+ *
+ * This file is part of eos-metrics-instrumentation.
+ *
+ * eos-metrics-instrumentation is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * eos-metrics-instrumentation is distributed in the hope that it will be
+ * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with eos-metrics-instrumentation.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include <glib.h>
+#include <stdio.h>
+
+#include "eins-network-id.h"
+
+#define PROC_IPV4_ROUTE_FILE "/proc/net/route"
+#define IPV4_ROUTE_REGEXP "^(?:\\S+\\s+){2}(?P<Gateway>[0-9A-F]{8})\\s+(?:\\S+\\s+){4}(?P<Mask>00000000)"
+
+/*
+ * greps the /proc/net/route file for the IPv4 default route. note that even
+ * on machines with multiple interfaces, the first listed 00000000 route will
+ * be the first one selected for outbound traffic, so we don't need to do any
+ * more complex searching/sorting here (verified in net/ipv4/fib_trie.c on
+ * kernel 4.10.5 - the trie is ordered by (prefix,tos,priority) so the highest
+ * preference default route will appear first)
+ */
+gchar *
+get_ipv4_default_gateway (void)
+{
+  g_autofree gchar *route = NULL;
+
+  if (!g_file_get_contents (PROC_IPV4_ROUTE_FILE, &route, NULL, NULL) || !route)
+    {
+      g_debug ("could not load " PROC_IPV4_ROUTE_FILE);
+      return NULL;
+    }
+
+  g_autoptr(GRegex) route_rx = g_regex_new (IPV4_ROUTE_REGEXP, G_REGEX_MULTILINE, 0, NULL);
```

Comment:
This doesn’t handle errors. If you’re confident no error could ever occur, it would still be good to have a `g_assert (route_rx != NULL)` just below.